tools/buildtools/schc.py

```diff
@@ -0,0 +1,923 @@
+    while i < len(tree):
+        expr = tree[i]
+
+        newTree.append(expr)
+
+        usedCurrentPostLabel = False
+        shouldAddPostLabel = True
+        if i + 1 < len(tree):
+            if tree[i + 1].token.tokenType == TokenType.LABEL:
+                # Re-use label if there's already one
+                currentPostLabel = tree[i + 1]
+                shouldAddPostLabel = False
+            else:
+                # dot (.) is used to ensure no name crashes with user-declared labels
+                labelToken = expr.token.newFromTokenType(TokenType.LABEL)
+                labelToken.tokenLiteral = f".{autoLabelName}.{depth}_{i}"
+                currentPostLabel = Expression(labelToken)
+
+        auxUsed = False
+        if expr.token.getProperties().isConditionalBranch:
+            if len(expr.left) == 0:
+                branchExpr = Expression(expr.token.newFromTokenType(TokenType.BRANCH), currentPostLabel.token)
+                expr.left.append(branchExpr)
+                if currentPostLabel == postLabel:
+                    usedLabel = True
+                else:
+                    usedCurrentPostLabel = True
+            else:
+                expr.left, auxUsed = normalizeTreeImpl(expr.left, currentPostLabel, depth+1, f"{autoLabelName}_left")
+                if currentPostLabel == postLabel:
+                    usedLabel = usedLabel or auxUsed
+
+            if len(expr.right) == 0:
+                branchExpr = Expression(expr.token.newFromTokenType(TokenType.BRANCH), currentPostLabel.token)
+                expr.right.append(branchExpr)
+                if currentPostLabel == postLabel:
+                    usedLabel = True
+                else:
+                    usedCurrentPostLabel = True
+            else:
+                expr.right, auxUsed = normalizeTreeImpl(expr.right, currentPostLabel, depth+1, f"{autoLabelName}_right")
+                if currentPostLabel == postLabel:
+                    usedLabel = usedLabel or auxUsed
+
+        if currentPostLabel != postLabel and (usedCurrentPostLabel or auxUsed) and shouldAddPostLabel:
+            newTree.append(currentPostLabel)
+
+        currentPostLabel = postLabel
+
+        i += 1
+
+    return newTree, usedLabel
+
+# Searches for `if_` and `else`s with empty bodies and inserts in them an unconditional branch to avoid compiler-induced fallthroughs
+def normalizeTree(tree: list[Expression]) -> list[Expression]:
+    postLabel = Expression(Token(TokenType.LABEL, ".autolabel.placeholder", "", -1, -1))
+
+    newTree, usedLabel = normalizeTreeImpl(tree, postLabel, 0, "_autolabel")
+
+    if usedLabel:
+        eprint("Warning: branching outside the script")
+        newTree.append(postLabel)
+    return newTree
+
+
+# For linearizing a tree
+@dataclasses.dataclass
+class LabeledExpression:
+    # Index relative to the parent list
+    index: int
+
+    token: Token
+    args: Token|None
+
+    # Label to jump into this command
+    labelName: str
```

Comment:
Btw I'm picking this line randomly but why not use the *de facto* standard for python comments on fields and variables,

```suggestion
    labelName: str
    """Label to jump into this command"""
```
iirc